### PR TITLE
Add Consent Page

### PR DESCRIPTION
**Pull request by mcygoogle**
Add consent page and template.
Add redirect operation in Authorization endpoint.
Add tests.

### CODE REVIEW

File: src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java [hunk 1]
```diff
@@ -118,5 +132,67 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
...
+                                        .build())
+                        .setRequestBody(
+                                OAuth2Request.RequestBody.newBuilder()
+                                        .setIsScoped(!scope.isEmpty())
+                                        .addAllScopes(scope)
+                                        .setResponseType(responseType)
+                                        .setRefreshable(responseType.equals("code"))
+                                        .setGrantType(
+                                                responseType.equals("code")
+                                                        ? "authorization_code": "implicit")
```
Comment (jstoops-at-google) [inline src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java:178]: This logic should be kept in sync with the similar code in validation. I would recommend a shared utility method that consistently gets the grant_type given the request's response_type.

File: src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java [hunk 2]
```diff
@@ -118,5 +132,67 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
...
+                    OAuth2Request.AuthorizationResponse.newBuilder()
+                            .setRedirectUri(redirectUri)
+                            .build());
+        } else {
+            oauth2RequestBuilder.setAuthorizationResponse(
+                    OAuth2Request.AuthorizationResponse.newBuilder()
+                            .setState(request.getParameter(OAuth2ParameterNames.STATE))
+                            .setRedirectUri(redirectUri)
+                            .build());
+        }
```
Comment (jstoops-at-google) [inline src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java:192]: nit: remove duplicate code here. You can also access submessage builders to make things a bit simpler.

```suggestion
        oauth2RequestBuilder.getAuthorizationResponseBuilder()
            .setRedirectUri(redirectUri);
        if (!Strings.isNullOrEmpty(request.getParameter(OAuth2ParameterNames.STATE))) {
            oauth2RequestBuilder.getAuthorizationResponseBuilder()
                .setState(request.getParameter(OAuth2ParameterNames.STATE));
        }
```

File: src/main/java/com/google/googleidentity/oauth2/endpoint/ConsentEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    protected void doPost(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException {
+        try {
+            toConsentPage(request, response);
+        } catch (TemplateException e) {
+            log.log(Level.INFO, "Display Consent Page Error!", e);
+        }
+    }
```
Comment (jstoops-at-google) [inline src/main/java/com/google/googleidentity/oauth2/endpoint/ConsentEndpoint.java:80]: Does this page need POST?

File: src/test/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpointTest.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        assertThat(requestBody.getGrantType()).isEqualTo("authorization_code");
+        assertThat(requestBody.getIsScoped()).isTrue();
+        assertThat(requestBody.getRefreshable()).isTrue();
+        assertThat(requestBody.getResponseType()).isEqualTo("code");
+        assertThat(requestBody.getScopesList()).containsExactlyElementsIn(CLIENT.getScopesList());
+
+        OAuth2Request.AuthorizationResponse authorizationResponse =
+                oauth2Request.getAuthorizationResponse();
+        assertThat(authorizationResponse.getState()).isEqualTo("111");
+        assertThat(authorizationResponse.getRedirectUri()).isEqualTo(REDIRECT_URI);
```
Comment (jstoops-at-google) [inline src/test/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpointTest.java:186]: I would strongly recommend using prototruth here. https://truth.dev/protobufs

Then, you can construct the entire expected proto, and a single assertion will give you an error message highlighting any diffs.

```
OAuth2Request.Builder expectedRequest = OAuth2Request.newBuilder().setClientId(...).setUserName(...);
expectedRequest.getRequestAuthBuilder().setGrantType(...)...

assertThat(oauth2Request).isEqualTo(expectedRequest.build());
```

File: src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java [hunk 5]
```diff
@@ -118,5 +132,64 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
         }
     }
 
+    /**
+     * Should be call after all validation in doGet function
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java:136]:
```suggestion
     * Should be called after all validations in doGet function.
```

File: src/main/java/com/google/googleidentity/oauth2/endpoint/ConsentEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    private Configuration configuration;
+
+    @Inject
+    public ConsentEndpoint() {
+    }
+
+    public void init() throws ServletException {
+
+        Version version = new Version("2.3.30");
```
Comment (weiminhuang-corp) [inline src/main/java/com/google/googleidentity/oauth2/endpoint/ConsentEndpoint.java:58]: Add a comment to explain where the value "2.3.30" comes from

**Pull request merged**